sidecar/src/service/runtime_info.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO-EK: Re-evaluate access levels for the structs and methods in this file after refactor,
+// before merging
+// TODO-EK: Better understand why apps and actions are conflated in this struct before merging. And,
+// if they should be separated.
+use crate::service::telemetry::AppInstance;
+use crate::service::{AppOrQueue, InstanceId, QueueId};
+use ddtelemetry::worker::{LifecycleAction, TelemetryActions};
+use futures::{
+    future::{self, join_all, Shared},
+    FutureExt,
+};
+use manual_future::{ManualFuture, ManualFutureCompleter};
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex, MutexGuard};
+use tracing::{debug, info};
+
+type AppMap = Arc<Mutex<HashMap<(String, String), Shared<ManualFuture<Option<AppInstance>>>>>>;
+
+/// `SharedAppManualFut` is a struct that contains a shared future of an `AppInstance` and its
+/// completer. The `app_future` is a shared future that may contain an `Option<AppInstance>`.
+/// The `completer` is used to complete the `app_future`.
+pub(crate) struct SharedAppManualFut {
+    pub(crate) app_future: Shared<ManualFuture<Option<AppInstance>>>,
+    pub(crate) completer: Option<ManualFutureCompleter<Option<AppInstance>>>,
+}
+
+/// `RuntimeInfo` is a struct that contains information about a runtime.
+/// It contains a map of apps and a map of app or actions.
+/// Each app is represented by a shared future that may contain an `Option<AppInstance>`.
+/// Each action is represented by an `AppOrQueue` enum.
+#[derive(Clone, Default)]
+pub(crate) struct RuntimeInfo {
+    pub(crate) apps: AppMap,
+    app_or_actions: Arc<Mutex<HashMap<QueueId, AppOrQueue>>>,
+    #[cfg(feature = "tracing")]
+    pub(crate) instance_id: InstanceId,
+}
+
+// TODO-EK: Can this be refactored more before merging? We may be able to encapsulate some of the
+// functionality in session_info to RuntimeInfo.
```

Comment:
Like for example?

Reply:
This TODO was just a reminder for myself while the PR was in draft. At the moment, I don't think this is possible in the short-term and isn't really blocking progress. I created a jira ticket and updated the TODO comment to reflect it.